packages/yarnpkg-core/sources/Project.ts

```diff
@@ -2122,8 +2122,64 @@ function applyVirtualResolutionMutations({
     }
   };
 
+  const devirtualizeSingleInstanceWorkspaces = () => {
+    const packagesWithInstanceCount = new Map<IdentHash, {virtualPackage: Package, descriptor: Descriptor, instanceCount: number}>();
+
+    // find any workspaces that only have one instance
+    for (const descriptor of allDescriptors.values()) {
+      const resolution = allResolutions.get(descriptor.descriptorHash);
+      if (!resolution) continue;
+
+      const virtualPackage = allPackages.get(resolution);
+      if (!virtualPackage) continue;
+
+      if (structUtils.isVirtualLocator(virtualPackage)) {
+        packagesWithInstanceCount.set(virtualPackage.identHash, {
```

Comment:
This doesn't work, you can have multiple versions of a single package that lists peer dependencies; since you're only keying on `identHash` (rather than `locatorHash` from the devirtualized locator), you will collapse them all into a single dict entry.

Reply:
Yes, but that is correct. I want to key on `identHash` to count the number of instances an `identHash` has. If there is more than one, it isn't affected.

Reply:
Unless you mean keying by `locatorHash` would result in additional packages being correctly devirtualized and affected by this change? 

Things that with `identHash` it currently misses?

Reply:
Yup, seems that was the case. With `identHash` it was more conservative (but still safe). `locatorHash` catches way more packages to be devirtualized.